# Chapter 8. 예외 처리

## 8.1. 코루틴의 예외 전파

- 코루틴 실행 도중 예외가 발생하면 예외가 발생한 코루틴은 취소되고 부모 코루틴으로 예외가 전파된다.
- 만약 부모 코루틴에서도 예외가 적절히 처리되지 않으면 부모 코루틴도 취소되고 루트 코루틴까지 반복될 수 있다.
  ![image](https://github.com/user-attachments/assets/53e738cf-61a7-46fc-a0b7-44e4ad7c739f)

## 8.2. 예외 전파 제한

### 8.2.1. Job 객체를 사용한 예외 전파 제한

- 새로운 Job 객체를 만들어 구조화를 깨고 싶은 코드를 연결
  ![image](https://github.com/user-attachments/assets/ad031f94-a22a-443a-8fb6-84f4903b5ec2)

- 하지만 이 방법을 사용하면 예외 전파뿐만 아니라 취소 전파도 제한함
- 그래서 큰 작업에 취소가 요청되더라도 취소 전파가 오지 않음

### 8.2.2. SupervisorJob 객체를 사용한 예외 전파 제한

- SupervisorJob 객체는 자식 코루틴으로부터 예외를 전파받지 않는 특수한 Job 객체

#### 8.2.2.2. 코루틴의 구조화를 깨지 않고 SupervisorJob 사용하기

- 구조화를 깨지 않고 SupervisorJob을 사용하기 위해서는 SupervisorJob의 인자로 부모 Job 객체를 넘기면 된다.
- 마지막에 supervisorJob.complete()을 실행해 명시적으로 완료 처리해야 한다.

#### 8.2.2.4. SupervisorJob을 사용할 때 흔히 하는 실수

- SupervisorJob 객체를 생설할 때 Job 계층 구조의 어떤 위치에 있어야 하는지 충분히 고민하고 사용해야 한다.
![image](https://github.com/user-attachments/assets/c86f22b5-5c30-49d4-a6ec-ceeda6a9867e)

### 8.2.3. supervisorScope를 사용한 예외 전파 제한

- supervisorScope 함수는 SupervisorJob 객체를 가진 CoroutineScope 객체를 생성함
- 이 SupervisorJob 객체는 supervisorScope 함수를 호출한 코루틴의 Job 객체를 부모로 가진다.
- 즉 복잡한 설정 없이도 구조화를 깨지 않고 예외 전파를 제한할 수 있다.
- 추가로 자식 코루틴이 모두 실행 완료되면 자동으로 완료 처리된다.


## 8.3. CoroutineExceptionHandler를 사용한 예외 처리

### 8.3.3. 처리되지 않은 예외만 처리하는 CoroutineExceptionHandler

- CoroutineExceptionHandler 객체는 처리되지 않는 예외만 처리한다.
- 만약 자식 코루틴이 부모 코루틴으로 예외를 전파하면 자식 코루틴에서는 예외가 처리된 것으로 보아 자식 코루틴에 설정된 CoroutineExceptionHandler 객체는 동작하지 않는다.

### 8.3.4. CoroutineExceptionHandler의 예외 처리 위치

- launch 코루틴 중 최상위에 있는 코루틴에 설정된 handler만 동작해 예외가 출력된다.
  ![image](https://github.com/user-attachments/assets/6415ea00-d3f7-41dc-a0a6-95210d41647d)

### 8.3.5. CoroutineExceptionHandler를 사용해야 하는 경우

- CoroutineExceptionHandler의 handlerException 함수가 호출될 때는 이미 해당 예외로 인해 코루틴이 완료된 상태이기 때문에 CoroutineExceptionHandler을 사용해 예외를 복구할 수 없다.
- 예외를 로깅하거나 오류 메시지를 표시하기 위해 공통으로 동작하는 예외 처리기를 설정해야 하는 경우 사용된다.

### 8.3.6. CoroutineExceptionHandler는 예외 전파를 제한하지 않는다

- handler는 예외가 마지막으로 처리되는 위치에서 예외를 처리할 뿐, 예외 전파를 제한하지 않는다.

## 8.4. try catch문을 사용한 예외 처리

### 8.4.1. try catch문을 사용해 코루틴 예외 처리하기

- try catch문을 사용하면 예외를 처리하고 예외 전파를 제한할 수 있다.

### 8.4.2. 코루틴 빌더 함수에 대한 try catch문은 코루틴의 예외를 잡지 못한다

- 코루틴 빌더 함수에 try catch문을 사용하면 코루틴에서 발생한 예외가 잡히지 않는다.
- 코루틴 빌더 함수를 try catch문으로 감싸면 빌더가 실행되는지만 확인할 뿐 빌더 안의 예외는 처리하지 못함

## 8.5. async의 예외 처리

### 8.5.1. async의 예외 노출

- async의 코루틴 실행 도중 예외가 발생하면 Deferred에 대한 await 호출 시 예외가 노출됨

### 8.5.2. async의 예외 전파

- await 함수 호출부에만 예외 처리를 할 게 아니라 async 코루틴 빌더 함수도 예외가 발생하면 부모 코루틴으로 예외를 전파하는데 이를 적절하게 처리해야 한다.
- 이를 위해 supervisorScope로 예외 전파를 제한해야 함

## 8.6. 전파되지 않는 예외

### 8.6.1. 전파되지 않는 CancellationException

- CancellationException은 부모 코루틴으로 전파되지 않는다
- 예외가 아니라 취소 확인용이라고 보면 됨

### 8.6.2. 코루틴 취소 시 사용되는 JobCancellationException

- Job 객체에 cancel 함수를 호출하면 CancellationException의 서브 클래스인 JobCancellationException을 발생시켜 코루틴을 취소시킨다.

### 8.6.3. withTimeOut 사용해 코루틴의 실행 시간 제한하기

- withTimeOut 함수는 작업이 주어진 시간 내에 완료되지 않으면 TimeoutCancellationException을 발생시킨다.
- try catch 문으로 해당 예외를 로그로 확인 가능
- withTimeOutOrNull을 사용하면 시간 초과 시 exception 대신 null을 반환함.

## 8.7. 요약

1. 애플리케이션은 다양한 예외 상황에 노출되며, 예외를 적절히 처리해 애플리케이션의 안정성을 확보할 수 있다.
2. 코루틴은 비동기 작업을 실행할 때 사용되기 때문에 애플리케이션의 안정성을 위해 예외 처리가 필수적이다.
3. 코루틴에서 발생한 예외는 부모 코루틴으로 전파되며, 적절히 처리되지 않으면 최상위 루트 코루틴까지 전파된다.
4. 예외를 전파받은 코루틴이 취소되면 해당 코루틴의 모든 자식 코루틴에 취소가 전파된다. 즉, 루트 코루틴이 예외로 취소되면 구조화된 모든 코루틴이 취소된다.
5. 새로운 루트 Job 객체를 통해 코루틴의 구조화를 깨서 코루틴의 예외 전파를 제한할 수 있다.
6. SupervisorJob 객체를 사용해 예외 전파를 제한할 수 있다. SupervisorJob 객체는 예외를 전파받지 않는 특수한 Job 객체이다.
7. SupervisorJob 객체는 예외를 전파받지 않지만 예외 정보는 전달받는다.
8. 예외가 전파되거나 예외 정보가 전달된 경우 해당 코루틴에서 예외가 처리된 것으로 본다.
9. CoroutineExceptionHandler 객체는 공통 예외 처리기로서 동작하며 이미 처리된 예외에 대해서는 동작하지 않는다. 즉, 예외가 마지막으로 전파되는 또는 전달되는 위치에 설정되지 않으면 동작하지 않는다.
10. CoroutineExceptionHandler 객체는 예외 전파를 제한하지 않는다.
11. 코루틴 내부에서 try catch 문을 사용해 예외를 처리할 수 있다.
12. 코루틴 빌더 함수에 대한 try catch문은 코루틴이 실행될 때 발생하는 예외를 잡지 못한다.
13. async 함수로 생성된 코루틴에서 발생한 예외는 await 호출 시 노출된다.
14. async 코루틴에서 발생한 예외 또한 부모 코루틴으로 전파된다.
15. CancellationException은 다른 예외와 달리 부모 코루틴으로 전파되지 않는다.
16. CancellationException이 전파되지 않는 이유는 CancellationException은 코루틴을 취소하기 위한 특별한 예외이기 때문이다. Job 객체에 cancel 함수를 호출하면 CancellationException의 서브 크랠스인 JobCancellationException이 발생해 코루틴이 취소된다.
17. withTimeOut 함수를 사용해 코루틴의 실행 시간을 제한할 수 있다. withTimeOut 함수는 실행 시간 초과 시 CancellationException의 서브 클래스인 TimeoutCancellationException을 발생시켜 해당 코루틴만 취소한다.
18. withTimeOutOrNull을 사용하면 실행 시간 초과 시 null이 반환되도록 할 수 있다.
